dialer/fastconnect.go

```diff
@@ -99,13 +113,23 @@ func (fcd *fastConnectDialer) connectAll(dialers []ProxyDialer) {
 		return
 	}
 	log.Debugf("Dialing all dialers in parallel %#v", dialers)
-	// Loop until we're connected
-	for len(fcd.connected.dialers) < 2 {
-		fcd.parallelDial(dialers)
-		// Add jitter to avoid thundering herd
-		time.Sleep(time.Duration(rand.Intn(4000)) * time.Millisecond)
+outerLoop:
+	for {
+		select {
+		case <-fcd.stopCh:
+			log.Debug("Stopping parallel dialing")
+			return
+		default:
+			// Loop until we're connected
+			if len(fcd.connected.dialers) < 2 {
+				fcd.parallelDial(dialers)
+				// Add jitter to avoid thundering herd
+				time.Sleep(time.Duration(rand.Intn(4000)) * time.Millisecond)
+			} else {
+				break outerLoop
+			}
+		}
 	}
```

Comment:
This could be simplified to:
```
for {
	// Loop until we're connected
	if len(fcd.connected.dialers) < 2 {
		fcd.parallelDial(dialers)
	} else {
		break
	}
	select {
	case <-fcd.stopCh:
		log.Debug("Stopping parallel dialing")
		return
	case time.After(time.Duration(rand.Intn(4000)) * time.Millisecond):
	}
}
```
But your way works too!

Reply:
I like it! Will change.